### PR TITLE
PR for issue #186: Transmit simulation time via FG UDP interface

### PR DESCRIPTION
**Pull request by simfinite**
Resolves feature request https://github.com/JSBSim-Team/jsbsim/issues/186

Please check whether it's ok to modify the documenting comment in net_fdm.hxx which is managed by the FlightGear project.

### CODE REVIEW

**Comment by bcoconni** [comment]
Thanks for the PR @simfinite .

Just a couple of comments:
1. The feature that you are submitting should be optional with the current behavior being the default (i.e. always send 1234567890). The fact that JSBSim does not use this parameter does not mean that FlightGear does not try to interpret it. I am concerned about the effect that this would have when using JSBSim as an external FDM to FlightGear.
2. Hard coded constants should be avoided. Please replace the `1E6` value by a parameter that can be modified by the user.

**Comment by simfinite** [comment]
Fair enough. I am assuming this should be implemented in terms of properties that can be acquired via the FGPropertyManager? 
If so, then I would appreciate some hints on how to implement this since I haven't worked with the property manager before. Is there a good example somewhere in the code on what it should look like? How are new properties defined? Where in the property tree should the new ones be placed?

**Comment by bcoconni** [comment]
You do not need to use properties but rather get information from the XML input files.

Let's say that the feature you are implementing will be requested with a syntax such that
```xml
<output name="localhost" type="FLIGHTGEAR" protocol="UDP" port="5550" rate="60">
  <time type="simulation" factor="1E+6"/>
</output>
```

So you need to check that there is an XML element `<time>` and that it has 2 attributes:
- `type` which should be set to `simulation` to indicate that the simulation time will be sent
- `factor` which provides the constant by which the time will be multiplied.

For that you need to add the inherited method `FGOutputFG::Load` that will set the boolean `useSimTime` and the double `factor` according to what is specified in the XML input:
```c++
FGOutputFG::Load(Element* el)
{
  // Check if there is a <time> element
  Element time_el = el->FindElement("time");
  if (time_el) {

    //Check if the attribute "type" is specified
    if (time_el->HasAttribute("type") && time_el->GetAttributeValue("type") == "simulation") {
      useSimTime = true;

      // Check if the attribute "factor" is specified
      if (time_el->HasAttribute("factor"))
        time_factor = time_el->GetAttributeValueAsNumber("factor");
    }
  }
}
```
Note that this (untested) code implies that `useSimTime` is initialized to `false` in the constructor and `factor` is initialized to `1E+6` in the constructor as well.

The code in `FGOutputFG::SocketDataFill` would then be something like
```c++
  if (useSimTime)
    net->cur_time = static_cast<uint32_t>(FDMExec->GetSimTime() * factor);
 else
    net->cur_time = (long int)1234567890;
```

This is just an example. Feel free to code the thing to your own taste.

**Comment by simfinite** [comment]
Thanks for the detailed example, I think I can take it from here.. just need to find a bit of free time.

**Comment by simfinite** [comment]
Ok, I finally got around to implementing this. I tested it with an accordingly modified ArduPlane SITL and it works as expected. We can now have a properly reliable lock-step mechanism to sync with JSBSim. If not activated via the tag `<time type="simulation">`, JSBSim exhibits the old behavior of sending 1234567890 as a dummy value.

File: src/input_output/FGOutputFG.cpp [hunk 1]
```diff
@@ -146,6 +147,35 @@ FGOutputFG::FGOutputFG(FGFDMExec* fdmex) :
...
+      outputOptions.useSimTime = true;
+    }
+
+    // Check if the attribute "resolution" is specified and set to a valid value
+    if (time_el->HasAttribute("resolution")) {
+      if (time_el->GetAttributeValueAsNumber("resolution") <= 1 &&
+          time_el->GetAttributeValueAsNumber("resolution") >= 1e-9) {
+        outputOptions.timeFactor = 1./time_el->GetAttributeValueAsNumber("resolution");
+      } else {
+        return false;
```
Comment (bcoconni) [inline src/input_output/FGOutputFG.cpp]: Ideally, an error message should be issued when such a condition is failed.

Comment (simfinite) [reply src/input_output/FGOutputFG.cpp]: Makes sense. How about:
```suggestion
      } else {
        cerr << "Not a valid time resolution (valid range is [1e-9, 1]): " << time_el->GetAttributeValueAsNumber("resolution") << endl;
        return false;
      }  
```

**Pull request merged**